exporter/redis.go

```diff
@@ -552,64 +550,52 @@ func (e *Exporter) handleMetricsCommandStats(addr string, alias string, fieldKey
 		return
 	}
 
-	e.metricsMtx.RLock()
-	defer e.metricsMtx.RUnlock()
-
 	cmd := splitKey[1]
-	e.metrics["commands_total"].WithLabelValues(addr, alias, cmd).Set(calls)
-	e.metrics["commands_duration_seconds_total"].WithLabelValues(addr, alias, cmd).Set(usecTotal / 1e6)
+	e.registerMetricValue(ch, "commands_total", calls, prometheus.CounterValue, cmd)
+	e.registerMetricValue(ch, "commands_duration_seconds_total", usecTotal/1e6, prometheus.CounterValue, cmd)
 }
 
-func (e *Exporter) handleMetricsReplication(addr string, alias string, fieldKey string, fieldValue string) bool {
-	e.metricsMtx.RLock()
-	defer e.metricsMtx.RUnlock()
-
+func (e *Exporter) handleMetricsReplication(ch chan<- prometheus.Metric, fieldKey string, fieldValue string) bool {
 	// only slaves have this field
 	if fieldKey == "master_link_status" {
 		if fieldValue == "up" {
-			e.metrics["master_link_up"].WithLabelValues(addr, alias).Set(1)
+			e.registerGaugeValue(ch, "master_link_up", 1)
 		} else {
-			e.metrics["master_link_up"].WithLabelValues(addr, alias).Set(0)
+			e.registerGaugeValue(ch, "master_link_up", 0)
 		}
 		return true
 	}
 
 	// not a slave, try extracting master metrics
-	if slaveOffset, slaveIp, slavePort, slaveState, lag, ok := parseConnectedSlaveString(fieldKey, fieldValue); ok {
-		e.metrics["connected_slave_offset"].WithLabelValues(
-			addr,
-			alias,
-			slaveIp,
-			slavePort,
-			slaveState,
-		).Set(slaveOffset)
-
-		if lag > -1 {
-			e.metrics["connected_slave_lag_seconds"].WithLabelValues(
-				addr,
-				alias,
-				slaveIp,
-				slavePort,
-				slaveState,
-			).Set(lag)
+	if slaveOffset, slaveIP, slavePort, slaveState, slaveLag, ok := parseConnectedSlaveString(fieldKey, fieldValue); ok {
+		e.registerGaugeValue(ch,
+			"connected_slave_offset",
```

Comment:
What's the unit for this?

Reply:
I think bytes. Rename it to `connected_slave_offset_bytes`?

Reply:
Yes, that sounds good.

Reply:
👍